### PR TITLE
Single-fetch typesafety

### LINKED ISSUE #9428: Single Fetch: useFetcher without generics arg triggers TS error `instantiation is excessively deep...`
### Reproduction

https://stackblitz.com/edit/remix-run-remix-k6ghpy?file=app%2Froutes%2F_index.tsx

### System Info

```shell
System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.18.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.3 - /usr/local/bin/npm
    pnpm: 8.15.6 - /usr/local/bin/pnpm
  npmPackages:
    @remix-run/dev: ^2.9.2 => 2.9.2 
    @remix-run/node: ^2.9.2 => 2.9.2 
    @remix-run/react: ^2.9.2 => 2.9.2 
    @remix-run/serve: ^2.9.2 => 2.9.2 
    vite: ^5.1.0 => 5.2.11
```

### Used Package Manager

npm

### Expected Behavior

Hi! For better typesafety we should give a loader/action function type parameter to the generic of `useFetcher`. But we sometimes don't do it, because of laziness or the action is too simple to bother, or nothing is done with the data. In these cases `useFetcher` (without generics arg) was just behaving normally, silently casting the `"data": any`.

### Actual Behavior

With SingleFetch the definition of `useFetcher` changed and it seems that we don't have default type parameter anymore:

```ts
  export function useFetcher<T extends Loader | Action>(
    opts?: Parameters<typeof useFetcherRR>[0]
  ): FetcherWithComponents<Serialize<T>>;
```
Would it be possible to have the `=  any` default parameter back?

### PR DESCRIPTION
**Pull request by pcattori**
Fixes #9428 #9514 #9649 #9714 #9776

## TODO

- [x] fix tests
- [x] `unstable_` prefix for `Future.singleFetch`
- [x] remove `Future` re-exports
- [x] Docs (`vite.config.ts` -> `declare module "@remix-run/server-runtime"`)
- [x] changeset

### CODE REVIEW

File: packages/remix-server-runtime/future.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export interface Future {}
```
Comment (brophdawg11) [inline packages/remix-server-runtime/future.ts:1]: Where does this get values added to it to trigger the ternary below?

```ts
// prettier-ignore
type SingleFetchEnabled =
  Future extends { singleFetch: infer T extends boolean } ? T : false
```

**Comment by punkpeye** [comment]
I think that the code in this PR has the same issue as described here https://github.com/remix-run/remix/issues/9776

**Comment by pcattori** [comment]
> I think that the code in this PR has the same issue as described here #9776

Turns out the new code with `Serialize<T>` helper doesn't have this issue, but just added some test cases and explicit handling for `ReadonlyArray` to be safe.

**Comment by pcattori** [comment]
@romansndlr don't get too excited, this is just a quick patch for improve types for single-fetch in Remix. The full typesafety story is coming soon in RRv7 😉 but yea this is the first step towards that

**Review by brophdawg11 — APPROVED** [review]
👌 great work!

**Comment by romansndlr** [comment]
> @romansndlr don't get too excited, this is just a quick patch for improved types for single-fetch in Remix, though yea that is pretty exciting on its own! The full typesafety story is coming soon in RRv7 😉 this is the first step towards that

For sure! I assumed as much... This is defiantly existing on it's own but also for what's to come!

File: docs/guides/single-fetch.md [hunk 2]
```diff
@@ -145,145 +145,120 @@ Without Single Fetch, any plain Javascript object returned from a `loader` or `a
...
-  | Set<Serializable>
-  | Promise<Serializable>;
+export default function Component() {
+  const data = useLoaderData<typeof loader>();
+  //    ^? { planet: string, date: Date, notSoRandom: undefined }
+}
 ```
 
-There are also client-side equivalents un `defineClientLoader`/`defineClientAction` that don't have the same return value restrictions because data returned from `clientLoader`/`clientAction` does not need to be serialized over the wire:
+Methods are also not serializable, so class instances get slimmed down to just their serializable properties:
```
Comment (sergiodxa) [inline docs/guides/single-fetch.md:199]: @pcattori Couldn't Remix call `classInstance.toJSON()`? If it's not defined then grab the serializable properties automatically

Comment (pcattori) [reply docs/guides/single-fetch.md:199]: It _could_ but that's what `json` is for

**Pull request merged**